### ISSUE TITLE
Store mask (instead of or in addition to prefix_length) in static lpm table entry

### ISSUE THREAD

**Issue by brunorijsman** [p4lang/p4c#2339]
Currently, static entries for longest prefix match keys only contain a prefix_length in the behavioral model JSON, for example:

          "entries" : [
              ....
              "match_key" : [
                {
                  "match_type" : "lpm",
                  "key" : "0x0a010200",
                  "prefix_length" : 24
                }
              ],

Which is produced by the following P4 program:

    header ipv4_header_t {
        bit<32> dst_addr;
        bit<8> ttl;
    }

...

    table ipv4_fib {
        key = {
            headers.ipv4_header.dst_addr: lpm;
        }
        actions = {
            act_miss;
            act_hit;
        }
        const default_action = act_miss();
        const entries = {
            32w0x0a010203 &&& 32w0xffffffff: act_hit(1);   // 10.1.2.3/32 -> 1
            32w0x0a010200 &&& 32w0xffffff00: act_hit(2);   // 10.1.2.0/24 -> 2
            32w0x0a010000 &&& 32w0xffff0000: act_hit(3);   // 10.1.0.0/16 -> 3
            32w0x0a000000 &&& 32w0xff000000: act_hit(4);   // 10.0.0.0/8 -> 4
        }
    }

However, to translate the prefix length into a bit mask we also need to know the length of the        key value element.

As a concrete example, if we know that the key value is "0x0a010200" and if we know that the prefix_length is 24, then that is not enough information to determine the mask (which is needed in the data plane). To determine that the mask is 0xffffff00 we also need to know that the field length is 32.

Determining length of the key value element using only the information in the behavioral model JSON turns out to be a remarkable complex and expensive operation.

It would have been much easier for the data path if the static lpm entry contained a mask instead of / in addition to prefix_length.

So, to summarize, I am asking that the behavior model is augmented as follows:

      "entries" : [
          ....
          "match_key" : [
            {
              "match_type" : "lpm",
              "key" : "0x0a010200",
              "prefix_length" : 24,
              "mask":  4294967040     // 0xffffff00
            }
          ],

(I have left the prefix_length in there for backwards compatibility, but we could remove it)

**Comment by mbudiu-vmw** [comment]
I think you should file this issue with the behavioral model repository; once they support this field it would be easy for the compiler to emit this.

**Comment by jafingerhut** [comment]
When you say: "Determining length of the key value element using only the information in the behavioral model JSON turns out to be a remarkable complex and expensive operation."

This information is available through several steps:

1. look at the list of fields in the lookup key defined for the table, which contains the header name
2. look up the header name in the "headers" section, which includes a "header_type" value
3. look up the header_type name in the "header_types" section, which defines the names and bit widths of all fields of the header.

Is that the complex and expensive operation you are referring to?  The result can certainly be cached in any kind of implementation that uses the BMv2 JSON for extracting such information.

**Comment by brunorijsman** [comment]
I am going to close the issue based on the following conversation on Slack:

Andy Fingerhut
I saw the other issue you created there, and responded to it,  Mihai's note that such an issue should be raised in the p4lang/behavioral-model repository is correct -- it cannot be changed in p4c alone without the behavioral-model code changing.  My comment is basically: I understand it could be made more convenient to access the key width for const entries keys, but the information is present in the BMv2 JSON file, and the way I described is exactly what the BMv2 simple_switch process must do today, and it probably caches the answer, too.

Bruno Rijsman
yes, that is my approach too: I gather the necessary information from other parts of the JSON and since that is a rather complex operation, I cache the result.  Do you think it is worthwhile raising the issue in p4lang/behavioral-model repository, or should I just drop it? (edited) 

Andy Fingerhut
The BMv2 JSON file format is kind of an implementation detail for simple_switch, and I know that a few other P4-related tools have chosen to use it for various purposes, too, and it is reasonably well documented, so it makes some sense that people would find other uses for it.  I personally would leave the decision to Antonin Bas, who has done more on simple_switch design/development work than anyone else, whether to add that kind of info to BMv2 JSON file, but I would not be surprised if he considered it unnecessarily redundant.
